Refactor parameter caching + add Zygote tests

SouthEndMusic opened the pull request:
Fixes #298
Fixes #301 

#289 can be a follow-up. Gradients w.r.t. u for `QuadraticSpline` are not broken anymore, but they are slow as mentioned in the issue.

The refactor includes:
- Caching parameters is opt-in, so users need to make the active decision to cache parameters. It is documented that this is not compatible with modifying `u` and `t`. We could if desired allow this with a dedicated interface which updates the parameters automatically. When parameters are not cached, they are computed on demand with the same functions.
- The `ReadOnlyArray` wrappers are removed, which simplifies the type parameters. I also noticed an inconsistency where I forgot to do it for integral inverse interpolations in the first place
- Also `safetycopy` as an option is removed again 

Zygote:
- I added Zygote tests for differentiation w.r.t. the interpolation input and w.r.t. the `u` of the data points for where that is supported (again). 
- I've added some examples in DataInterpolationsChainRulesCoreExt.jl of how to speed up gradients w.r.t. u with analytical expressions. Speeding up more interpolation types this way could be a follow up

SouthEndMusic commented:
How can I fix the Zygote compat?

ChrisRackauckas commented on Project.toml:
```diff
@@ -33,14 +32,14 @@ LinearAlgebra = "1.10"
 StableRNGs = "1"
 Symbolics = "5.29"
 Test = "1"
+Zygote = "^0"
```
```suggestion
Zygote = "0.6"
```

ChrisRackauckas commented:
You can only compat weakdeps or deps. So added as a weakdep if we want the compat.

SouthEndMusic commented:
I've added tests for I think everything that was compatible with Zygote before. I've added some examples in `DataInterpolationsChainRulesCoreExt.jl` of how to speed up gradients w.r.t. u with analytical expressions. Speeding up more interpolation types this way could be a follow up
